core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -393,11 +393,13 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
             }
             accessControl.checkCanInsertIntoTable(session.toSecurityContext(), targetTable);
 
-            TableMetadata tableMetadata = metadata.getTableMetadata(session, targetTableHandle.get());
-
-            List<ColumnMetadata> columns = tableMetadata.getColumns().stream()
+            List<ColumnMetadata> columns = metadata.getColumnHandles(session, targetTableHandle.get())
+                    .values().stream()
+                    .map(columnHandle -> metadata.getColumnMetadata(session, targetTableHandle.get(), columnHandle))
```

Comment:
it's  O(columns)  calls into connector metadata now

Reply:
Don't we use `CachingJdbcClient` to reduce amount of calls?

Reply:
Yes. That is correct. My understanding behind that I liked it at first is that it do not actually reach the real metadata, as it could be possibly to take that information just of of table handle.

Hence my comment: https://github.com/trinodb/trino/pull/6745#discussion_r565964405

Anyway, I think we need a dedicated method in SPI. This is just hacking.

Reply:
> Don't we use CachingJdbcClient to reduce amount of calls?

@ssheikin `JdbcMetadata#getColumnMetadata` is O(1) and does not even reach `CachingJdbcClient`
But this is generic code, and needs to work great for every connector

Reply:
> Does it usually call any metadata service to get column metadata? Please check HiveMetadata and JdbcMetadata

No I didn't find any implementation which calls metadata service.
In most cases it has simple implementation:
```
    @Override
    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
    {
        return ((JdbcColumnHandle) columnHandle).getColumnMetadata();
    }
```

>  it do not actually reach the real metadata, as it could be possibly to take that information just of of table handle.

Seems it is the case.